include/h2o/balancer.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017 Justin Zhu
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#ifndef h2o__balancer_h
+#define h2o__balancer_h
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "h2o/socketpool.h"
+#include "yoml.h"
+
+/* function for configure per-target extra data when parsing configuration. node = NULL for default */
+typedef int (*h2o_balancer_per_target_conf_parser)(yoml_t *node, void **data, yoml_t **errnode, char **errstr);
+
+/* function for parsing overall configuration of a load balancer */
+typedef int (*h2o_balancer_overall_conf_parser)(yoml_t *node, void **data, yoml_t **errnode, char **errstr);
```

Comment:
Considering the complexity, I think that it would be preferable to have unit tests (i.e. tests written in C under t/00unit/lib) that confirms the behavior of each load balancer.

That would mean that:
* each balancer would have its own constructor
* each balancer should be programatically configurable (either as part of the arguments passed to the constructor or as separate functions)
* a separate function that accepts a YOML node and either constructs or configures a balancer needs to exist

I like the fact that you have written end-to-end tests already, and I think we need them regardless of if we add unit tests. However I also believe that it is hard to test every corner of the balancer logic using end-to-end tests, and that it would be easier for us to confirm and maintain compatibility in future if we have unit tests here.